DOXYGEN: Further doc updates in high priority files

b-gent opened the pull request:
Adding, editing, updating doxygen comments in header files.

HTML output available at: https://b-gent.github.io/PR2535/d6/d8f/class_common_1_1_config_manager.html


<!---
Thank you for contributing to ScummVM. Please read the following carefully before submitting your Pull Request.

Make sure your individual commits follow the guidelines found in the ScummVM Wiki: https://wiki.scummvm.org/index.php?title=Commit_Guidelines. If they're not please edit them before submitting the Pull Request.

Commits and Pull Requests should use the following template:

```
SUBSYSTEM: Short (50 chars or less) summary of changes

More detailed explanatory text, if necessary.  Wrap it to about 72
characters or so.  In some contexts, the first line is treated as the
subject of an email and the rest of the text as the body.  The blank
line separating the summary from the body is critical (unless you omit
the body entirely); tools like rebase can get confused if you run the
two together.

Write your commit message in the present tense: "Fix bug" and not "Fixed
bug."  This convention matches up with commit messages generated by
commands like git merge and git revert.

Further paragraphs come after blank lines.

- Bullet points are okay, too

- Typically a hyphen or asterisk is used for the bullet, preceded by a
 single space, with blank lines in between, but conventions vary here

- Use a hanging indent
```
--->

b-gent commented:
The last commit added the rest of missing doxygen in str.h. I won't add any more files to this PR and will open a new one instead, it's already a lot to review.
HTML output is updated here: https://b-gent.github.io/PR2535/d8/dbd/group__common__str.html

Please use the suggestion option to correct me in the given lines, I'm sure there will be quite a few. My approach was to add as much as possible, even where I really wasn't sure.

I will rebase tomorrow, the merge conflict in config-manager looks pretty bad.
@sev-

mgerhardy commented on common/array.h:
```diff
@@ -221,16 +226,19 @@ class Array {
 		return _storage[idx];
 	}
 
+    /** Return a pointer to the underlying memory serving as element storage. */
```
These an a few of the following look like whitespace errors

b-gent replied:
thanks, I'll make sure to align with function declaration

b-gent replied:
They all seem well-aligned actually, you can check it for yourself in the raw source view of the file -> https://github.com/scummvm/scummvm/blob/3bd1b124f233d2a66cecf532de1f60e6f1d6140f/common/array.h

Unless you meant something else by 'whitespace error'? The whitespace between the full stop and doxygen comment closing is intentional and used consistently everywhere as part of doxygen style, for better readability.

b-gent commented:
Rebased. It wasn't as bad as it seemed - [this commit](https://github.com/scummvm/scummvm/commit/465807f165a2d0a4f3f8fb73686ce52db677bbaa) introduced some serious changes in line spacing so git showed a lot of changed lines but content was not the issue. So the only addition coming from master was [this](https://github.com/scummvm/scummvm/commit/ee4799c51aee8096968f9088f85b915dfac9c109).

b-gent commented on common/config-manager.h:
```diff
@@ -60,38 +60,44 @@ class ConfigManager : public Singleton<ConfigManager> {
-		bool tryGetVal(const String &key, String &out) const { return _entries.tryGetVal(key, out); }
+		String        &getVal(const String &key) { return _entries.getVal(key); } /*!< Retrieve the value of a @p key. */
+		const String  &getVal(const String &key) const { return _entries.getVal(key); } /*!< @overload */
+         /**
```
@henke37  you added this recently, can you let me know what is the right description?

henke37 replied:
It retrieves the value if the key exists and leaves the referenced variable unchanged if the key does not exist.

b-gent replied:
thanks

criezy commented on common/array.h:
```diff
@@ -82,13 +82,16 @@ class Array {
 
+	/**
+	 * @todo Doc required
+	 */
```
Construct an array as a copy of the given @p array.

I think this should be explicit enough for developers reading the doc. But if we wanted more details we could say:
Construct an array with the same size as the given @p array and whose elements are copies of the elements in @p array.

criezy commented on common/array.h:
```diff
@@ -107,7 +110,7 @@ class Array {
 
 	/**
-	 * Constructs an array using list initialization.
+	 * Construct an array using list initialization.
```
Maybe we should add an example to clarify what list initialisation is?
For example
`Common::Array<int> myArray = {1, 7, 42};`
constructs an array with 3 elements whose values are 1, 7, and 42 respectively.

Also since currently C++11 support can be disabled when running configure, we want to indicate that this constructor is only available when C++11 support is enabled.

criezy commented on common/array.h:
```diff
@@ -221,16 +226,19 @@ class Array {
 		return _storage[idx];
 	}
 
+    /** @todo Doc required. */
```
Assign the given @p array to this array.

criezy commented on common/array.h:
```diff
@@ -244,6 +252,7 @@ class Array {
 	}
 
 #ifdef USE_CXX11
+    /** @todo Doc required. */
```
Assign the given array to this array using the C++11 move semantic.

criezy commented on common/array.h:
```diff
@@ -261,29 +270,34 @@ class Array {
 		_capacity = 0;
 	}
 
+    /** Erase the element at @p pos position. */
```
...and return an iterator pointing to the next element in the array.

criezy commented on common/array.h:
```diff
@@ -261,29 +270,34 @@ class Array {
 		return (_size == 0);
 	}
 
+    /** Check if two arrays are of the same size. */
```
It does not just check that the arrays have the same size, but also that they have the same content.
So in short: `Check if two arrays are identical`.

criezy commented on common/array.h:
```diff
@@ -296,26 +310,32 @@ class Array {
 		return !(*this == other);
 	}
 
+    /** Return a pointer to the underlying memory where the array begins. */
```
Again, I would not mention `underlying memory` here.
I would suggest: `Return an iterator pointing to the first element in the array.`

criezy commented on common/array.h:
```diff
@@ -296,26 +310,32 @@ class Array {
 		return _storage;
 	}
 
+    /** Return a pointer to the underlying memory where the array ends. */
```
Return an iterator pointing past the last element in the array.

criezy commented on common/array.h:
```diff
@@ -296,26 +310,32 @@ class Array {
 		return _storage + _size;
 	}
 
+    /** Return a pointer to the underlying memory where the array begins. */
```
Return a const iterator pointing to the first element in the array.

criezy commented on common/array.h:
```diff
@@ -296,26 +310,32 @@ class Array {
 		return _storage;
 	}
 
+    /** Return a pointer to the underlying memory where the array ends. */
```
Return a const iterator pointing past the last element in the array.

criezy commented on common/array.h:
```diff
@@ -296,26 +310,32 @@ class Array {
 		return _storage + _size;
 	}
 
+    /** Copy array data to a new place in memory. */
```
It might copy the array data to a new place in memory, but this is a side effect, and may not happen.

I would suggest: `Reserve enough memory in the array so that it can store at least the given number of elements. The current content of the array is not modified.`

criezy commented on common/array.h:
```diff
@@ -330,13 +350,15 @@ class Array {
 		_size = newSize;
 	}
 
+    /** @todo Doc required. */
```
Assign to this array the elements between the given iterators (from @p first included to @p last excluded) from another array.

criezy reviewed (changes requested):
I made a partial review of this pull request and it needs some changes.
I will continue the review tomorrow.

criezy commented on common/rect.h:
```diff
@@ -109,91 +143,104 @@ static inline Point operator*(double multiplier, const Point &p) { return Point(
 	Rect(int16 w, int16 h) : top(0), left(0), bottom(h), right(w) {}
+	/**
+     * Create a rectangle between the points defined with @p x1, @p y1, @p x2, and @p y2.
+	 * Additionally, check whether it is a valid rectangle.
```
The `Additionally, check whether it is a valid rectangle.` is not really correct. It gives the impression that it is safe to pass invalid coordinates. But such a check in a `assert` actually causes a crash if the check fails. It is used to detect incorrect code.

So it would be more correct to say that it create a rectangle with top-left corner at the given x1 and y1 positions, and the bottom-right corner at the given x2 and y2 positions. The x2 value has to be equal or bigger than the x1 value and the y2 value has to be equal or bigger than the y1 value.

b-gent replied:
Changed to 
```
	/**
     * Create a rectangle with top-left corner at the given (x1,y1) position
	 * and the bottom-right corner at the (x2,y2) position.
	 * 
	 * The @p x2 value must be greater or equal @p x1 and @p y2 must be greater or equal @p y1.
     */
```

criezy commented on common/rect.h:
```diff
@@ -109,91 +143,104 @@ static inline Point operator*(double multiplier, const Point &p) { return Point(
 
 	Rect() : top(0), left(0), bottom(0), right(0) {}
+	/**
+     * Create a rectangle with width @p w and height @p h.
```
Maybe mention `a rectangle with the top-left corner at position (0, 0) and the given width @p w and height @p h`?

criezy commented on common/rect.h:
```diff
@@ -109,91 +143,104 @@ static inline Point operator*(double multiplier, const Point &p) { return Point(
 		assert(isValidRect());
 	}
+	/**
+     * Create a rectangle with width @p w and height @p h.
```
Where does that come from?
This is a comparison operator that checks if two rectangles are identical. It returns true if the two rectangles are the same and false otherwise.

criezy commented on common/rect.h:
```diff
@@ -109,91 +143,104 @@ static inline Point operator*(double multiplier, const Point &p) { return Point(
+     */
 	bool operator==(const Rect &rhs) const { return equals(rhs); }
+    /**
+     * Create a rectangle with width @p w and height @p h.
```
And another comparison operator that checks if two rectangles are different. It returns true if the two rectangles are different, and false if they are the same.

criezy commented on common/rect.h:
```diff
@@ -109,91 +143,104 @@ static inline Point operator*(double multiplier, const Point &p) { return Point(
 
-	int16 width() const { return right - left; }
-	int16 height() const { return bottom - top; }
+	int16 width() const { return right - left; }  /*!< Compute the width of a rectangle. */
```
I would not say `compute` here, but just `return`.

It's an implementation detail that the rectangle stores internally the left and right position rather than the left position and the width. The two would be equivalent for the developer using a rectangle.

criezy commented on common/rect.h:
```diff
@@ -109,91 +143,104 @@ static inline Point operator*(double multiplier, const Point &p) { return Point(
-	int16 width() const { return right - left; }
-	int16 height() const { return bottom - top; }
+	int16 width() const { return right - left; }  /*!< Compute the width of a rectangle. */
+	int16 height() const { return bottom - top; } /*!< Compute the height of a rectangle. */
```
And same comment as above for the width.

criezy commented on common/rect.h:
```diff
@@ -226,6 +273,9 @@ struct Rect {
 	}
 
+	/**
+	 * Clip this reactangle with another rectangle @p r.
```
Typo in `reactangle` ;-)

criezy commented on common/rect.h:
```diff
@@ -243,41 +293,62 @@ struct Rect {
 	}
 
+   	/**
+	 * @todo Doc required.
```
Return true if the rectangle is empty (the width or the height is 0) or is invalid (negative width or height). Return false if this is a non-empty valid rectangle.

criezy commented on common/rect.h:
```diff
@@ -243,41 +293,62 @@ struct Rect {
 	}
 
+	/**
+	 * @todo Doc required.
```
Move the rectangle by the given delta x and y values.

criezy commented on common/savefile.h:
```diff
@@ -54,48 +54,89 @@ typedef SeekableReadStream InSaveFile;
 
 public:
-	OutSaveFile(WriteStream *w);
+	OutSaveFile(WriteStream *w); /*!< Create a WriteStream instance @p w. */
```
It doesn't create a WriteStream instance. It creates an `OutSaveFile` that uses the given `WriteStream` to actually write the data.

Technically, since `OutSaveFile` is itself a `WriteStream`, it does create a `WriteStream`, but not @p w. The `w` WriteStream already exists.

criezy reviewed (commented):
I reviewed some more changes.
I have not checked the changes to `str.h` yet. I will do that tomorrow.

b-gent commented:
@criezy  thanks so much ! I will start implementing the comments tomorrow and let you know when done.

criezy commented:
There were significant changes to `common/str.h` this morning, so I will wait until that is resolved to review the changes in that file. One possibility would be to drop the changes to `common/str.h` from this PR and document it, as well as `common/base-str.h`, in a new PR.

b-gent commented:
> There were significant changes to `common/str.h` this morning, so I will wait until that is resolved to review the changes in that file. One possibility would be to drop the changes to `common/str.h` from this PR and document it, as well as `common/base-str.h`, in a new PR.

Yes, I can see there are quite a few changes there and resolving the conflict will be very error-prone because in most cases I modify the same lines. I guess we will go for the safer solution of moving this file to a separate PR.

b-gent commented on common/array.h:
```diff
@@ -221,16 +226,19 @@ class Array {
 
 	// TODO: insert, remove, ...
 
+	/** Return a pointer to the underlying memory serving as element storage. */
```
Thanks for the explanation, these little bits of information really help to understand the code.

b-gent commented on common/list.h:
```diff
@@ -131,48 +131,49 @@ class List {
 	}
 
-	/** Returns a reference to the first element of the list. */
+	/** Return a reference to the first element of the list. */
```
If 'returning' is the only thing that the function does, I think it's better to keep it as a ``@brief`` (we are using autobrief so all these descriptions are essentially a ``@brief`` element). Honestly, I'm not even sure if it would be possible to have a ``@return`` without a brief first. And duplicating the same thing in both brief and return does not make much sense.

That said, ``@return`` or ``@retval`` is of course a great thing to have in more complex functions that can return various values.

criezy replied:
I am not familiar with `@brief` (we decided to turn it off at work) and forgot it exists. But I think that what you wrote makes sense.

b-gent commented on common/list.h:
```diff
@@ -131,48 +131,49 @@ class List {
 		return static_cast<Node *>(_anchor._prev)->_data;
 	}
 
+    /** @todo Doc required. */
```
So ``Assign a given @p list to this list`` would be correct?

b-gent commented on common/rect.h:
```diff
@@ -44,35 +44,69 @@ namespace Common {
+     */
 	Point operator-(const Point &delta) const { return Point(x - delta.x, y - delta.y); }
+	/**
+     * Create a point by dividing the x and y arguments by the @p delta value.
```
yes, I overlooked a copy-paste issue, thanks

b-gent commented on common/rect.h:
```diff
@@ -44,35 +44,69 @@ namespace Common {
+     */
 	bool  operator!=(const Point &p)    const { return x != p.x || y != p.y; }
+	/**
+     * Create a point by adding the @p delta value to the x and y arguments.
```
Changed to
`Create a point by adding the @p delta value to a point.`
(and similar in all of these)

b-gent commented:
@criezy  all your comments are addressed now

b-gent commented:
I have reverted ``str.h`` to the state of master but Github is still showing that I changed things in that file. But that's not right, for example it is showing that I added `#include "common/base-str.h"` , while this has been recently introduced to master through another PR.

criezy commented:
I merged the commit manually to double check what was happening with the `str.h` file.
